File: backend/app/schemas/document.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, computed_field


class DocumentLinks(BaseModel):
    """What a document is filed against. All independent, all optional.

    A signed contract belongs to a deal, an NDA to the person *and* the company
    it was signed with — so these are lists, and a document can sit under
    several records at once without being copied.
    """

    contact_ids: list[UUID] = []
    organization_ids: list[UUID] = []
    deal_ids: list[UUID] = []
    project_ids: list[UUID] = []


class DocumentRead(DocumentLinks):
    id: UUID
    title: str
    description: str | None
    tags: list[str]
    format: str
    size: int
    storage_key: str
    preview_key: str | None = Field(exclude=True)
    created_at: datetime
    updated_at: datetime

    @computed_field  # type: ignore[prop-decorator]
    @property
    def has_preview(self) -> bool:
        return self.preview_key is not None

    model_config = {"from_attributes": True}


class DocumentUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    tags: list[str] | None = None
    # Sent as a whole list, like the project router's link fields: the list
    # replaces what was there, and an empty list detaches everything.
    contact_ids: list[UUID] | None = None
    organization_ids: list[UUID] | None = None
    deal_ids: list[UUID] | None = None
    project_ids: list[UUID] | None = None

File: frontend/lib/models/document.dart
class Document {
  const Document({
    required this.id,
    required this.title,
    this.description,
    required this.tags,
    required this.format,
    required this.size,
    required this.storageKey,
    required this.hasPreview,
    this.contactIds = const [],
    this.organizationIds = const [],
    this.dealIds = const [],
    this.projectIds = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String title;
  final String? description;
  final List<String> tags;
  final String format;
  final int size;
  final String storageKey;
  final bool hasPreview;

  // What the document is filed against. A signed contract belongs to a deal,
  // an NDA to the person *and* the company — so these are lists, and the same
  // file can sit under several records without being uploaded twice.
  final List<String> contactIds;
  final List<String> organizationIds;
  final List<String> dealIds;
  final List<String> projectIds;

  /// How many records this document is filed against, all kinds together.
  int get linkCount =>
      contactIds.length +
      organizationIds.length +
      dealIds.length +
      projectIds.length;

  final DateTime createdAt;
  final DateTime updatedAt;

  factory Document.fromJson(Map<String, dynamic> json) => Document(
        id: json['id'] as String,
        title: json['title'] as String,
        description: json['description'] as String?,
        tags: (json['tags'] as List<dynamic>).cast<String>(),
        format: json['format'] as String,
        size: json['size'] as int,
        storageKey: json['storage_key'] as String,
        hasPreview: json['has_preview'] as bool? ?? false,
        contactIds: (json['contact_ids'] as List<dynamic>?)?.cast<String>() ?? [],
        organizationIds:
            (json['organization_ids'] as List<dynamic>?)?.cast<String>() ?? [],
        dealIds: (json['deal_ids'] as List<dynamic>?)?.cast<String>() ?? [],
        projectIds: (json['project_ids'] as List<dynamic>?)?.cast<String>() ?? [],
        createdAt: DateTime.parse(json['created_at'] as String),
        updatedAt: DateTime.parse(json['updated_at'] as String),
      );
}
